x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/pagerduty/IncidentEvent.java

```diff
@@ -93,46 +94,96 @@ public int hashCode() {
-                            payload.toXContent(builder, params);
-                            builder.endObject();
-                        }
-                        if (contexts != null && contexts.length > 0) {
-                            builder.startArray(Fields.CONTEXTS.getPreferredName());
-                            for (IncidentEventContext context : contexts) {
-                                context.toXContent(builder, params);
-                            }
-                            builder.endArray();
-                        }
-                        return builder;
-                    }
-                })
+                .setHeader("Accept", ACCEPT_HEADER)
+                .jsonBody((b, p) -> buildAPIXContent(b, p, serviceKey, payload, watchId))
                 .build();
     }
 
+    public XContentBuilder buildAPIXContent(XContentBuilder builder, Params params, String serviceKey,
+                                            Payload payload, String watchId) throws IOException {
+        builder.field(Fields.ROUTING_KEY.getPreferredName(), serviceKey);
+        builder.field(Fields.EVENT_ACTION.getPreferredName(), eventType);
+        if (incidentKey != null) {
+            builder.field(Fields.DEDUP_KEY.getPreferredName(), incidentKey);
+        }
+
+        builder.startObject(Fields.PAYLOAD.getPreferredName());
+        {
+            builder.field(Fields.SUMMARY.getPreferredName(), description);
+
+            if (attachPayload && payload != null) {
+                builder.startObject(Fields.CUSTOM_DETAILS.getPreferredName());
+                {
+                    builder.field(Fields.PAYLOAD.getPreferredName(), payload, params);
+                }
+                builder.endObject();
+            }
+
+            if (watchId != null) {
+                builder.field(Fields.SOURCE.getPreferredName(), watchId);
+            } else {
+                builder.field(Fields.SOURCE.getPreferredName(), "watcher");
+            }
+            // TODO externalize this into something user editable
+            builder.field(Fields.SEVERITY.getPreferredName(), "critical");
+        }
+        builder.endObject();
+
+        if (client != null) {
+            builder.field(Fields.CLIENT.getPreferredName(), client);
+        }
+        if (clientUrl != null) {
+            builder.field(Fields.CLIENT_URL.getPreferredName(), clientUrl);
+        }
+
+        if (contexts != null && contexts.length > 0) {
+            toXContentV2Contexts(builder, params, contexts);
+        }
+
+        return builder;
+    }
+
+    /**
+     * Turns the V1 API contexts into 2 distinct lists, images and links. The V2 API has separated these out into 2 top level fields.
+     */
+    private void toXContentV2Contexts(XContentBuilder builder, ToXContent.Params params,
+                                      IncidentEventContext[] contexts) throws IOException {
+        // contexts can be either links or images, and the v2 api needs them separate
```

Comment:
how about making this more readable (at least to me :-)

```
        Map<IncidentEventContext.Type, List<IncidentEventContext>> groupByType =
            Arrays.stream(contexts).collect(Collectors.groupingBy(ctx -> ctx.type));

        List<IncidentEventContext> linkContexts = groupByType.get(IncidentEventContext.Type.LINK);
        if (linkContexts != null && linkContexts.isEmpty() == false) {
            builder.startArray(Fields.LINKS.getPreferredName());
            for (IncidentEventContext context : linkContexts) {
                context.toXContent(builder, params);
            }
            builder.endArray();
        }
        List<IncidentEventContext> imagesContexts = groupByType.get(IncidentEventContext.Type.IMAGE);
        if (imagesContexts != null && imagesContexts.isEmpty() == false) {
            builder.startArray(Fields.IMAGES.getPreferredName());
            for (IncidentEventContext context : imagesContexts) {
                context.toXContent(builder, params);
            }
            builder.endArray();
        }

```

Reply:
dang i like those streams.. I still am a streaming noob. ty.